src/Model.js

```diff
@@ -471,13 +499,57 @@ class Model {
    * @method _makeObservable
    */
   _makeObservable (initialAttributes) {
-     const { defaultAttributes } = this
+    const { defaultAttributes } = this
 
-     extendObservable(this, {
-       ...defaultAttributes,
-       ...initialAttributes
-     })
-   }
+    extendObservable(this, {
+      ...defaultAttributes,
+      ...initialAttributes
+    })
+
+    this._listenForChanges()
+  }
+
+  /**
+   * sets up a reaction for each top-level attribute so we can compare
+   * values after each mutation and keep track of dirty attr states
+   * if an attr is different than the last snapshot, add it to the
+   * _dirtyAttributes set
+   * if it's the same as the last snapshot, make sure it's _not_ in the
+   * _dirtyAttributes set
+   * @method _listenForChanges
+   */
+  _listenForChanges () {
+    this._disposers = Object.keys(this.attributes).map((attr) => {
+      return reaction(() => this.attributes[attr], (value) => {
+        const previousValue = this.previousSnapshot.attributes[attr]
+        if (isEqual(previousValue, value)) {
+          this._dirtyAttributes.delete(attr)
+        } else if (isObject(value)) { // handles Objects and Arrays
+          // clear out any dirty attrs that start with this attr prefix
+          // then we can reset them if they are still (or newly) dirty
+          Array.from(this._dirtyAttributes).forEach((path) => {
+            if (path.indexOf(`${attr}.`) === 0) {
+              this._dirtyAttributes.delete(path)
+            }
+          })
+          diff(previousValue, value).forEach((property) => {
+            this._dirtyAttributes.add(`${attr}.${property}`)
+          })
+        } else {
+          this._dirtyAttributes.add(attr)
+        }
+      })
+    })
+  }
+
+  /**
+   * call this when destroying an object to make sure that we clean up
+   * any event listeners and don't leak memory
+   * @method dispose
+   */
+  dispose () {
```

Comment:
@jordanstephens, where does `dispose` get called? Do we want to call this in the `destroy` method?

Reply:
it doesn't get called right now. i just added it for completeness. we could certainly call it in `destroy` and we should probably think about other places we will want to call it in portal. evicting from cache for instance.